prusti-interface/src/specs/encoder.rs

```diff
@@ -0,0 +1,108 @@
+use prusti_rustc_interface::data_structures::fx::{FxHashMap, FxIndexSet};
+use prusti_rustc_interface::hir::def_id::{CrateNum, DefId, DefIndex};
+use prusti_rustc_interface::middle::mir::interpret::AllocId;
+use prusti_rustc_interface::middle::ty::{self, PredicateKind, Ty};
+use prusti_rustc_interface::middle::ty::codec::TyEncoder;
+use prusti_rustc_interface::middle::ty::TyCtxt;
+use prusti_rustc_interface::serialize::{Encodable, Encoder, opaque};
+
+pub struct DefSpecsEncoder<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    opaque: opaque::MemEncoder,
+    type_shorthands: FxHashMap<Ty<'tcx>, usize>,
+    predicate_shorthands: FxHashMap<PredicateKind<'tcx>, usize>,
+    interpret_allocs: FxIndexSet<AllocId>,
+}
+
+impl<'tcx> DefSpecsEncoder<'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>) -> Self {
+        DefSpecsEncoder {
+            tcx,
+            opaque: opaque::MemEncoder::new(),
+            type_shorthands: Default::default(),
+            predicate_shorthands: Default::default(),
+            interpret_allocs: Default::default(),
+        }
+    }
+
+    pub fn into_inner(self) -> Vec<u8> {
+        self.opaque.finish()
+    }
+}
+
+macro_rules! encoder_methods {
```

Comment:
I assume some of this is also from `rustc`? If so, add a comment similar to the decoder one here, also with a permalink.